### PR TITLE
Fix item has no attribute error

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
Add `type: ignore` to Ignore the below errors:
- Item "None" of "Optional[CompilationState]" has no attribute "prefix"
- "Type[T]" has no attribute "from_json"
- Module has no attribute (It seems like a bug in mypy) https://stackoverflow.com/questions/60734086/mypy-gives-an-error-while-importing-submodule-module-has-no-attribute. sqlalchemy has a attribute `create_engine`, but mypy will raise an error `Module "sqlalchemy" has no attribute "create_engine"`

Fix mypy errors like `Item "None" of "Optional[Interface]" has no attribute "inputs"`
```
flytekit/core/reference_entity.py:114: error: Item "None" of "Optional[Interface]" has no attribute "inputs"
flytekit/core/reference_entity.py:124: error: Item "None" of "Optional[Interface]" has no attribute "outputs"
flytekit/core/reference_entity.py:136: error: Item "None" of "Optional[Any]" has no attribute "outputs"
flytekit/core/reference_entity.py:137: error: Item "None" of "Optional[Interface]" has no attribute "outputs"
flytekit/core/reference_entity.py:158: error: Item "None" of "Optional[Any]" has no attribute "inputs"
flytekit/core/reference_entity.py:159: error: Item "None" of "Optional[Interface]" has no attribute "inputs"
flytekit/core/reference_entity.py:166: error: Item "VoidPromise" of "Union[VoidPromise, Any, Any]" has no attribute "literals"
flytekit/core/reference_entity.py:167: error: Item "None" of "Optional[Interface]" has no attribute "outputs"
```

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
Pass mypy checks
`make test | grep "attribute"`

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1447

## Follow-up issue
_NA_

### CODE REVIEW

File: flytekit/core/condition.py [hunk 1]
```diff
@@ -176,8 +176,8 @@ def end_branch(self) -> Union[Condition, Promise]:
         """
         ctx = FlyteContextManager.current_context()
         # Let us mark the execution state as complete
-        ctx.execution_state.branch_complete()
-        if self._last_case:
+        ctx.execution_state.branch_complete()  # type: ignore
+        if self._last_case and self._selected_case:
```
Comment (wild-endeavor) [inline flytekit/core/condition.py:180]: was this a bug?!?

**Review by wild-endeavor — APPROVED** [review]
just one comment

File: flytekit/core/condition.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/core/condition.py:180]: `self._selected_case` could be `None`. We have to check it, otherwise mypy will raise an error.

File: flytekit/core/condition.py [hunk 2]
```diff
@@ -176,8 +176,8 @@ def end_branch(self) -> Union[Condition, Promise]:
...
             # We have completed the conditional section, lets pop off the branch context
             FlyteContextManager.pop_context()
             if self._selected_case.output_promise is None and self._selected_case.err is None:
```
Comment (pingsutw) [inline flytekit/core/condition.py:183]: If we don't check it, this line might throw an exception

Comment (wild-endeavor) [reply flytekit/core/condition.py:183]: oh right

**Comment by wild-endeavor** [comment]
feel free to ignore integration tests for now, we're still working on them

**Comment by wild-endeavor** [comment]
thank you for doing all this btw, this has been bothering me forever.

**Review by wild-endeavor — APPROVED** [review]
